### PR TITLE
RSDK-5425 - Fix the vectornav IMU

### PR DESCRIPTION
**Pull request by nfranczak**
This PR refactors the vectornav imu driver. 
```json
    {
      "name": "movement_sensor-1",
      "namespace": "rdk",
      "type": "movement_sensor",
      "model": "imu-vectornav",
      "attributes": {
        "chip_select_pin": "0",
        "polling_freq_hz": 200,
        "spi_baud_rate": 9600,
        "spi_bus": "0"
      }
    }
```

### CODE REVIEW

File: components/movementsensor/imuvectornav/imu.go
Comment (susmitaSanyal) [inline components/movementsensor/imuvectornav/imu.go:32]: The name for `CSPin` can be a little more descriptive, GPIOPin. We are not really selecting any chip here

File: components/movementsensor/imuvectornav/imu.go
Comment (susmitaSanyal) [inline components/movementsensor/imuvectornav/imu.go:119]: add an info level message here to declare that we are setting the default baudRate to 9600. Also let's make a global constant variable `DefaultBaudRate` and set it to 9600 and use it here

File: components/movementsensor/imuvectornav/imu.go
Comment (susmitaSanyal) [inline components/movementsensor/imuvectornav/imu.go:122]: Why is it called smp? Can we rename this variable to something that makes more sense?

File: components/movementsensor/imuvectornav/imu.go
Comment (susmitaSanyal) [inline components/movementsensor/imuvectornav/imu.go:117]: oooof, `newVectorNav` is doing so many things. It's hard to keep up. I think we can break it into five different functions: 
1. Configuration Parsing and Setup 
2. Register Reading 
3. IMU Configuration 
4. Acceleration Bias Compensation 
5. Polling Setup 

And call these five functions above in `newVectorNav`

File: components/movementsensor/imuvectornav/imu.go
Comment (susmitaSanyal) [inline components/movementsensor/imuvectornav/imu.go:208]: What is s?

File: components/movementsensor/imuvectornav/imu.go
Comment (susmitaSanyal) [inline components/movementsensor/imuvectornav/imu.go:305]: NICE!!! Thanks for converting them to m/s^2

**Review by susmitaSanyal — CHANGES_REQUESTED** [review]

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go:122]: will rename to `pollFreq`

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go:208]: seems to be the hertz version of the smp aka pollFreq. I can rename s to pollHertz. Thoughts?

**Comment by nfranczak** [comment]
@susmitaSanyal should I add `Reconfigure` as well?

File: components/movementsensor/imuvectornav/imu.go
Comment (randhid) [reply components/movementsensor/imuvectornav/imu.go]: No we are, we're selecting a channel from the chip select, it cannot be any GPIO pin, renaming it would be wrong.

**Comment by randhid** [comment]
> @susmitaSanyal should I add `Reconfigure` as well?

No do not add reconfigure, reconfigure is poison.

File: components/movementsensor/imuvectornav/imu.go
Comment (randhid) [inline components/movementsensor/imuvectornav/imu.go:70]: why this change?

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go]: I will revert the name

File: components/movementsensor/imuvectornav/imu.go
Comment (randhid) [inline components/movementsensor/imuvectornav/imu.go:148]: I actually prefer the flatter version of the constructor; I do not like setting things from within the helper functions because of potential locks with a spi communication protocol, which the Linux kernel fdopen does not ensure thread safety with. So please make it flatter again.

**Review by randhid — CHANGES_REQUESTED** [review]
Please also test with a raspberry pi once you've addressed all comments and take a screenshot of the device working. Nice job!

File: components/movementsensor/imuvectornav/imu.go
Comment (randhid) [inline components/movementsensor/imuvectornav/imu.go:192]:
```suggestion
func (vn *vectornav) configureVectorNav(ctx context.Context) error {
```
but I'd rather this be in the constructor.

Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go:192]: I am slightly confused by this comment. Above you say that you prefer a "flatter" version of constructor, but here you give me a naming suggestion. Do you want me to make the constructor flatter while also keeping this helper function?

Comment (randhid) [reply components/movementsensor/imuvectornav/imu.go:192]: Two things are true:
- the name of this function does not currently nicely describe what it is doing.
- I prefer flat constructors, but I know susmita does not, so that's up to you. For me the happy path is not having structs populated by helper functions, or the stoppable workers started from within a helper function (rather, have the stoppable workers consume a helper function after everything in the imu struct has been set. I don't find the extra layers helpful from a debugging perspective.

File: components/movementsensor/imuvectornav/imu.go
Comment (randhid) [inline components/movementsensor/imuvectornav/imu.go:74]:
```suggestion
	csPin   string
```
You missed one.

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go:70]: `compensateAccelBias` and `compensateDVBias` expect the secondary argument to be of type `uint`. This way we do not have to cast them when the function is called and can instead simply do `v.polling`

Comment (randhid) [reply components/movementsensor/imuvectornav/imu.go:70]: I see, this is a good change we could even go earlier and cast it in the config

File: components/movementsensor/imuvectornav/imu.go
Comment (randhid) [reply components/movementsensor/imuvectornav/imu.go:192]: Just to be clear, we can add helpers, but I do prefer that helpers are designed to be isolated and testable and have a reason.

Comment (randhid) [reply components/movementsensor/imuvectornav/imu.go:192]: You could also, in general, approach tickets which need a refactor like so:
- address the bug first, add a new pr refactoring
- prefactor the code to make it more readable (limit such tickets to 1-2 days), address the bug in a follow up pr.

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go:70]: oo nice! I will do this.

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go:192]: > Just to be clear, we can add helpers, but I do prefer that helpers are designed to be isolated and testable and have a reason.

Them being testable is a good point.
I do not know how any of the helper functions could be tested since we would need to interact with the imu itself.
For the reason of the helpers not being testable I will revert back the constructor so that it is completely flat.

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go:120]: Per this conversation, https://github.com/viamrobotics/rdk/pull/4445#discussion_r1799757422.
Specifically that none of these helpers would be testable I think it makes most sense to keep the constructor structured as is.

File: components/movementsensor/imuvectornav/imu.go
Comment (randhid) [inline components/movementsensor/imuvectornav/imu.go:215]: This is actually in units of seconds since it's 1/frequency

**Comment by nfranczak** [comment]
video of the component working
https://github.com/user-attachments/assets/5bbc3f3c-6649-4d5a-93d3-23adf588f1d1

**Comment by nfranczak** [comment]
vectornav at time = 1
Screenshot 2024-10-14 at 1 19 42 PM

**Comment by nfranczak** [comment]
vectornav at time = 2

I flipped it so that it was upside down
Screenshot 2024-10-14 at 1 19 58 PM

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go:215]: will rename to `pollPerSecond`

File: components/movementsensor/imuvectornav/imu.go
Comment (randhid) [inline components/movementsensor/imuvectornav/imu.go:28]:
```suggestion
const defaultBaudRate = 9600
```
no need for this to be exported.

File: components/movementsensor/imuvectornav/imu.go
Comment (randhid) [inline components/movementsensor/imuvectornav/imu.go:202]: Hmm, a potential race with constructing and closing with a bad configuration is possible now since we're using the struct's `v.polling` rather than a new variable that's set to the value. However, this was more of a concern when we used to read and write spi directly to memory, which will require re-implementing this driver entirely if we go back to using that as an underlying mechanism for the imu, I don't know how worried to be about this.

Also could we do 
```suggestion
	if err = v.compensateAccelBias(ctx, v.polling); err != nil {
```
for brevity?

**Review by randhid — APPROVED** [review]
Thanks for testing, looks good to me now, but please wait for Olivia and Susmita to review as well.

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go]: > Hmm, a potential race [...] is possible now since we're using the struct's v.polling rather than a new variable that's set to the value. 

If there is a possibility of a data race since we are now using v.polling lets just go back to using the new variable instead.

File: components/movementsensor/imuvectornav/imu.go
Comment (oliviamiller) [inline components/movementsensor/imuvectornav/imu.go:35]: adding comment here that CSPIn is either 0 or 1 would be helpful.

File: components/movementsensor/imuvectornav/imu.go
Comment (oliviamiller) [inline components/movementsensor/imuvectornav/imu.go:201]: typo mesurament -> measurement

File: components/movementsensor/imuvectornav/imu.go
Comment (oliviamiller) [inline components/movementsensor/imuvectornav/imu.go:206]: polling frequency is alway greater than zero because we set it to 200 at line 126 or conf value

File: components/movementsensor/imuvectornav/imu.go
Comment (oliviamiller) [inline components/movementsensor/imuvectornav/imu.go:206]: This is fine with me but the previous comment said that we were optionally polling, now we will be polling all the time, is this what we want the behavior to be?

**Review by oliviamiller — APPROVED** [review]
lgtm just one question

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go:206]: I believe so.
Previously, it was "optional polling" on the condition that we had a valid frequency to poll at, i.e. it was not a negative value.
Now that we set by default to have a valid polling frequency it being optional is moot.

File: components/movementsensor/imuvectornav/imu.go
Comment (randhid) [inline components/movementsensor/imuvectornav/imu.go:35]:
```suggestion
	CSPin    string `json:"chip_select_pin"` // This value will be either 1 or 0 for linux's default spi bus
```

File: components/movementsensor/imuvectornav/imu.go
Comment (susmitaSanyal) [reply components/movementsensor/imuvectornav/imu.go]: Then the comment needs to reflect and explain what chip_select_pin means

Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go]: @susmitaSanyal comment has been updated

**Review by susmitaSanyal — APPROVED** [review]
LGTM, approving with the assumption that it passes hardware test

File: components/movementsensor/imuvectornav/imu.go
Comment (nfranczak) [inline components/movementsensor/imuvectornav/imu.go:33]: @susmitaSanyal @oliviamiller @randhid 

is there a specific reason for why `BaudRate` and `Pfreq` are pointers to an int?

I am asking since if I choose to leave them as empty in the attributes section of the config, the constructor freaks out. 
However, if I change them to be of type `int` the problem goes away.

Comment (randhid) [reply components/movementsensor/imuvectornav/imu.go:33]: It was a trick to check if a config field was present in the validator back in the day when we didn't have good json schemas for the configs. We can make them not pointers for now to avoid nil checks.

Comment (nfranczak) [reply components/movementsensor/imuvectornav/imu.go:33]: hooray!

**Pull request merged**
